AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt

```diff
@@ -29,246 +29,404 @@ import com.google.android.material.card.MaterialCardView
-
-							if (messageAttachment.filename.endsWith(".ogg")) {
-								var file = File(ctx.cacheDir, "audio.ogg")
-								file.deleteOnExit()
-								Http.simpleDownload(url, file)
-								url = file.absolutePath
-							}
-
-							currentPlayer?.apply {
-								setDataSource(url)
-								setOnPreparedListener {
-									seekTo((sliderView.progress.div(500f) * duration).toInt())
-									Utils.mainThread.post { updatePlaying() }
-									duration = it.duration.toLong()
-								}
-								setOnCompletionListener { player ->
-									playing = false
-									player.seekTo(0)
-									Utils.mainThread.post { buttonView.background = rewindIcon }
-								}
-								currentPlayerUnsubscribe = {
-									playing = false
-									Utils.mainThread.post { updatePlaying() }
-								}
-								onPauseListener = {
-									playing = false
-									Utils.mainThread.post { updatePlaying() }
-								}
-								prepare()
-								isPrepared = true
-								preparing = false
-							}
-						}
-					} else {
-						updatePlaying()
-					}
-				}
-
-				addView(buttonView)
-				addView(progressView)
-				addView(sliderView)
-			})
-		}
-
-		patcher.after<StoreMessages>("handleChannelSelected", Long::class.javaPrimitiveType!!) {
-			currentPlayerUnsubscribe?.invoke()
-			currentPlayerUnsubscribe = null
-			onPauseListener = null
-		}
-
-		patcher.after<AppActivity>("onCreate", Bundle::class.java) {
-			currentPlayerUnsubscribe?.invoke()
-			currentPlayerUnsubscribe = null
-			onPauseListener = null
-		}
-
-		patcher.after<AppActivity>("onPause") {
-			onPauseListener?.invoke()
-		}
-	}
-
-	override fun stop(context: Context) {
-		patcher.unpatchAll()
-	}
-}
+    fun deleteOggCacheFiles(cacheDir: File, oggCacheMap: Map<String, File>) {
+        val oggFiles = cacheDir.listFiles { file ->
+            file.parentFile == cacheDir &&
+            file.name.matches(Regex("^audio_-?\\d+\\.ogg$"))
+        } ?: emptyArray()
+        val filesToDelete = oggFiles.filter { file -> !oggCacheMap.values.contains(file) }
+        for (file in filesToDelete) {
+            file.delete()
+        }
+    }
+
+    override fun start(context: Context) {
+        deleteOggCacheFiles(cacheDir = context.cacheDir, oggCacheMap = oggFileCache)
+
+        patcher.after<WidgetChatListAdapterItemAttachment>(
+            "configureFileData",
+            MessageAttachment::class.java,
+            MessageRenderContext::class.java
+        ) {
+            val messageAttachment = it.args[0] as MessageAttachment
+            val root = WidgetChatListAdapterItemAttachment.`access$getBinding$p`(this).root as ConstraintLayout
+            val card = root.findViewById<MaterialCardView>(attachmentCardId)
+            val ctx = root.context
+
+            card.findViewById<MaterialCardView>(playerBarId)?.let { card.removeView(it) }
+            val loadingBarId = playerBarId + 1
+            card.findViewById<ProgressBar>(loadingBarId)?.let { card.removeView(it) }
```

Comment:
Don't remove views, instead set their visibility to GONE (and back to visible once needed)